### PR TITLE
Fix Date metricid cache consistency under concurrent use

### PR DESCRIPTION
**Pull request by rtm0**
### Describe Your Changes

Fix Date metricid cache consistency under concurrent use.
When one goroutine calls Has() and does not find the cache entry in the immutable map it will acquire a lock and check the mutable map. And it is possible that before that lock is acquired, the entry is moved from the mutable map to the immutable map by another goroutine causing a cache miss.

The fix is to check the immutable map again once the lock is acquired. 

### Checklist

The following checks are **mandatory**:

- [x ] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### CODE REVIEW

File: lib/storage/storage_test.go [hunk 1]
```diff
@@ -156,6 +157,30 @@ func testDateMetricIDCache(c *dateMetricIDCache, concurrent bool) error {
...
+	dmc := newDateMetricIDCache()
+	var wg sync.WaitGroup
+	for i := range concurrency {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for id := uint64(i * numMetrics); id < uint64((i+1)*numMetrics); id++ {
+				dmc.Set(generation, date, id)
+				if !dmc.Has(generation, date, id) {
+					panic(fmt.Errorf("dmc.Has(metricID=%d): unexpected cache miss after adding the entry to cache", id))
```
Comment (f41gh7) [inline lib/storage/storage_test.go:176]: Could you please replace panic with `t.Fatalf` ? It requires to change function signature `(t *testing.T)`

Comment (rtm0) [reply lib/storage/storage_test.go:176]: Please allow me to explain before I change anything. So `t.Fatalf` shouldn't be called from goroutines and, when running `make check-all`, `go vet` would complain (even if I add that param to the goroutine):

```
go vet ./lib/...
# github.com/VictoriaMetrics/VictoriaMetrics/lib/storage
# [github.com/VictoriaMetrics/VictoriaMetrics/lib/storage]
lib/storage/storage_test.go:176:6: call to (*testing.T).Fatalf from a non-test goroutine
```

Instead, the error would need to be propataged to the main test goroutine which'd complicate the code. And using `panic` somewhat simplifies it.

Let me know what you think. Thanks!

**Comment by rtm0** [comment]
> Overall looks good. Could you please mention changes at changelog https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/docs/CHANGELOG.md

Done. Please note that the fix was for `vmsingle` but I guess it will also be merged into the cluster version as well (not sure yet how this happens and who does this), thus I mentioned both vmsingle and cluster versions in the changelog.

**Review by f41gh7 — APPROVED** [review]
LGTM

**Comment by f41gh7** [comment]
> Done. Please note that the fix was for `vmsingle` but I guess it will also be merged into the cluster version as well (not sure yet how this happens and who does this), thus I mentioned both vmsingle and cluster versions in the changelog.

 It's completely fine. We're back-porting changes from single to cluster and enterprise branches on our own.

**Comment by f41gh7** [comment]
Thanks for contribution!

**Pull request merged**